# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

In [1]:
# import numpy and pandas
import pandas as pd
import numpy as np
from scipy.stats import trim_mean, mode, skew, gaussian_kde, pearsonr, spearmanr, beta
from statsmodels.stats.weightstats import ztest as ztest

from scipy.stats import ttest_ind, norm, t
from scipy.stats import f_oneway
from scipy.stats import sem

In [2]:
from pathlib import Path

import numpy as np
import pandas as pd
from scipy import stats
from statsmodels.stats.proportion import proportions_ztest

# Look for the data folder in the supplied project structure.
possible_data_dirs = [
    Path("../data"),      # original lab repository structure
   Path("data"),         # current working directory structure
    Path("../../data"),   # alternative structure
]   

DATA_DIR = next((p for p in possible_data_dirs if p.exists()), None)

if DATA_DIR is None:
    raise FileNotFoundError(
        "Could not find the Data folder. Please check the project folder structure."
    )

print(f"Using data folder: {DATA_DIR.resolve()}")




Using data folder: /Users/theophiluskobinasarpey/Documents/DSML_SEP2026 | IRONHACK/Week 4/Day 2/lab-hypothesis-testing-en/data


### Challenge 1 - Exploring the Data

In this challenge, we will examine all salaries of employees of the City of Chicago. We will start by loading the dataset and examining its contents

In [4]:
# Run this code:
# Load the employee salary dataset.
salaries = pd.read_csv(
    DATA_DIR / "Current_Employee_Names__Salaries__and_Position_Titles.csv"
)

# Display the first five rows to understand the structure.
salaries.head()


,Name,Job Titles,Department,Full or Part-Time,Salary or Hourly,Typical Hours,Annual Salary,Hourly Rate
0,"AARON, JEFFERY M",SERGEANT,POLICE,F,Salary,NaN,101442.0,NaN
1,"AARON, KARINA",POLICE OFFICER (ASSIGNED AS DETECTIVE),POLICE,F,Salary,NaN,94122.0,NaN
2,"AARON, KIMBERLEI R",CHIEF CONTRACT EXPEDITER,GENERAL SERVICES,F,Salary,NaN,101592.0,NaN
3,"ABAD JR, VICENTE M",CIVIL ENGINEER IV,WATER MGMNT,F,Salary,NaN,110064.0,NaN
4,"ABASCAL, REECE E",TRAFFIC CONTROL AIDE-HOURLY,OEMC,P,Hourly,20.0,NaN,19.86


Examine the `salaries` dataset using the `head` function below.

In [5]:
# Your code here
# Display the first five rows to understand the structure.
salaries.head()


,Name,Job Titles,Department,Full or Part-Time,Salary or Hourly,Typical Hours,Annual Salary,Hourly Rate
0,"AARON, JEFFERY M",SERGEANT,POLICE,F,Salary,NaN,101442.0,NaN
1,"AARON, KARINA",POLICE OFFICER (ASSIGNED AS DETECTIVE),POLICE,F,Salary,NaN,94122.0,NaN
2,"AARON, KIMBERLEI R",CHIEF CONTRACT EXPEDITER,GENERAL SERVICES,F,Salary,NaN,101592.0,NaN
3,"ABAD JR, VICENTE M",CIVIL ENGINEER IV,WATER MGMNT,F,Salary,NaN,110064.0,NaN
4,"ABASCAL, REECE E",TRAFFIC CONTROL AIDE-HOURLY,OEMC,P,Hourly,20.0,NaN,19.86


In [ ]:
# Display the last five rows to understand the structure and confirm the number of rows or observations.
salaries.tail()

,Name,Job Titles,Department,Full or Part-Time,Salary or Hourly,Typical Hours,Annual Salary,Hourly Rate
33178,"ZYLINSKA, KATARZYNA",POLICE OFFICER,POLICE,F,Salary,NaN,72510.0,NaN
33179,"ZYMANTAS, LAURA C",POLICE OFFICER,POLICE,F,Salary,NaN,48078.0,NaN
33180,"ZYMANTAS, MARK E",POLICE OFFICER,POLICE,F,Salary,NaN,90024.0,NaN
33181,"ZYRKOWSKI, CARLO E",POLICE OFFICER,POLICE,F,Salary,NaN,93354.0,NaN
33182,"ZYSKOWSKI, DARIUSZ",CHIEF DATA BASE ANALYST,DoIT,F,Salary,NaN,115932.0,NaN


In [6]:
# Inspect the dimensions and column names.
print("Number of rows:", len(salaries))
print("Number of columns:", salaries.shape[1])
print("\nColumns:")
print(salaries.columns.tolist())

# Count salaried versus hourly employees.
print("\nEmployee type counts:")
print(salaries["Salary or Hourly"].value_counts())


Number of rows: 33183
Number of columns: 8

Columns:
['Name', 'Job Titles', 'Department', 'Full or Part-Time', 'Salary or Hourly', 'Typical Hours', 'Annual Salary', 'Hourly Rate']

Employee type counts:
Salary or Hourly
Salary    25161
Hourly     8022
Name: count, dtype: int64


### Challenge 1 result interpretation
The dataset contains 33,183 employees and 8 columns.
There are:
- 25,161 salaried employees
- 8,022 hourly employees

For the remaining challenges, we restrict the analysis to the 8,022 hourly workers.

### Challenge 2 - Hypothesis Tests

In this section of the lab, we will test whether the hourly wage of all hourly workers is significantly different from $30/hr. Import the correct one sample test function from scipy and perform the hypothesis test for a 95% two sided confidence interval.

### Question
Is the average hourly wage of Chicago's hourly workers significantly different from $30/hour?
Because we are comparing one sample mean with a fixed reference value, the appropriate test is a one-sample t-test.
Hypotheses
- H₀: μ = $30/hour
- H₁: μ ≠ $30/hour
This is a two-sided test because the question asks whether the mean is different, rather than specifically higher or lower.
Significance level
We use:
α = 0.05
Test assumptions
The one-sample t-test assumes:
1. Observations are independent.
2. The underlying population is approximately normal, or the sample is large enough for the Central Limit Theorem to make the sampling distribution of the mean approximately normal.
3. The measurement is quantitative.
Here the sample contains 8,022 hourly workers, so the large sample size makes the t-based inference very robust to moderate departures from normality.

In [8]:
# Your code here
# Keep only hourly workers and their observed hourly rate.
hourly_workers = salaries.loc[
    salaries["Salary or Hourly"].eq("Hourly"),
    "Hourly Rate"
].dropna()

# Basic descriptive statistics.
print(f"Number of hourly workers: {len(hourly_workers):,}")
print(f"Sample mean: ${hourly_workers.mean():.2f}/hour")
print(f"Sample standard deviation: ${hourly_workers.std(ddof=1):.2f}/hour")


Number of hourly workers: 8,022
Sample mean: $32.79/hour
Sample standard deviation: $12.11/hour


In [9]:
# Perform a two-sided one-sample t-test against $30/hour.
t_test_result = stats.ttest_1samp(
    hourly_workers,
    popmean=30,
    alternative="two-sided"
)

print(f"t-statistic = {t_test_result.statistic:.6f}")
print(f"p-value     = {t_test_result.pvalue:.6e}")


t-statistic = 20.619806
p-value     = 4.323024e-92


### Challenge 2 — Interpretation
The test gives:
- Sample mean: $32.79/hour
- t-statistic: 20.6198
- p-value: 4.323e-92
- α: 0.05
Because the p-value is far below 0.05, we reject H₀.

Conclusion
There is extremely strong statistical evidence that the mean hourly wage of hourly workers is different from $30/hour.
The observed sample mean is about $32.79/hour, which is higher than $30/hour.

# Challenge 3 - Constructing Confidence Intervals

While testing our hypothesis is a great way to gather empirical evidence for accepting or rejecting the hypothesis, another way to gather evidence is by creating a confidence interval. A confidence interval gives us information about the true mean of the population. So for a 95% confidence interval, we are 95% sure that the mean of the population is within the confidence interval. 
).

To read more about confidence intervals, click [here](https://en.wikipedia.org/wiki/Confidence_interval).


In the cell below, we will construct a 95% confidence interval for the mean hourly wage of all hourly workers. 

The confidence interval is computed in SciPy using the `t.interval` function. You can read more about this function [here](https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.stats.t.html).

To compute the confidence interval of the hourly wage, use the 0.95 for the confidence level, number of rows - 1 for degrees of freedom, the mean of the sample for the location parameter and the standard error for the scale. The standard error can be computed using [this](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.sem.html) function in SciPy.

### Challenge 3 — Constructing a 95% Confidence Interval
A hypothesis test gives us evidence about a specific value. A confidence interval gives us a range of plausible values for the population mean.
For a 95% confidence interval we use:
mean ± t-critical × standard error
where:
- mean is the sample mean.
- t-critical comes from the t-distribution.
- standard error = sample standard deviation / √n.
- Degrees of freedom = n - 1.
SciPy's t.interval() performs this calculation directly.

In [10]:
# Your code here
# Calculate the sample size, mean, and standard error.
n = len(hourly_workers)
sample_mean = hourly_workers.mean()
standard_error = stats.sem(hourly_workers)

# Construct the 95% confidence interval using the t-distribution.
confidence_interval = stats.t.interval(
    confidence=0.95,
    df=n - 1,
    loc=sample_mean,
    scale=standard_error
)

print(f"Sample size: {n:,}")
print(f"Sample mean: ${sample_mean:.4f}/hour")
print(f"Standard error: ${standard_error:.6f}")
print(
    f"95% confidence interval: "
    f"${confidence_interval[0]:.2f} to ${confidence_interval[1]:.2f} per hour"
)



Sample size: 8,022
Sample mean: $32.7886/hour
Standard error: $0.135237
95% confidence interval: $32.52 to $33.05 per hour


### Challenge 3 — Interpretation
The 95% confidence interval is approximately $32.52 to $33.05 per hour.
The benchmark of $30/hour is not inside this interval. This agrees with the hypothesis test: at the 5% significance level, we reject the null hypothesis that the population mean is $30/hour.

In practical terms, the data places the mean hourly wage around $32.79/hour, with a fairly narrow interval because the sample is very large.

# Challenge 4 - Hypothesis Tests of Proportions

Another type of one sample test is a hypothesis test of proportions. In this test, we examine whether the proportion of a group in our sample is significantly different than a fraction. 

You can read more about one sample proportion tests [here](http://sphweb.bumc.bu.edu/otlt/MPH-Modules/BS/SAS/SAS6-CategoricalData/SAS6-CategoricalData2.html).

In the cell below, use the `proportions_ztest` function from `statsmodels` to perform a hypothesis test that will determine whether the number of hourly workers in the City of Chicago is significantly different from 25% at the 95% confidence level.

### Challenge 4 — Hypothesis Test of a Proportion
Now we test a different type of parameter: a population proportion.
Question
Is the proportion of hourly workers in the City of Chicago significantly different from 25%?
The appropriate test is a one-sample proportion z-test.
Hypotheses
Let p be the proportion of all employees who are hourly workers.
- H₀: p = 0.25
- H₁: p ≠ 0.25
Again, this is a two-sided test.
Significance level
α = 0.05

In [11]:
# Your code here
# Count hourly workers and all employees.
hourly_count = (salaries["Salary or Hourly"] == "Hourly").sum()
total_count = len(salaries)

# Calculate the observed sample proportion.
observed_proportion = hourly_count / total_count

print(f"Hourly workers: {hourly_count:,}")
print(f"Total employees: {total_count:,}")
print(f"Observed proportion: {observed_proportion:.4%}")


Hourly workers: 8,022
Total employees: 33,183
Observed proportion: 24.1750%


In [12]:
# Test whether the hourly-worker proportion differs from 25%.
z_stat, p_value = proportions_ztest(
    count=hourly_count,
    nobs=total_count,
    value=0.25,
    alternative="two-sided"
)

print(f"z-statistic = {z_stat:.6f}")
print(f"p-value     = {p_value:.6f}")


z-statistic = -3.509996
p-value     = 0.000448


### Challenge 4 — Interpretation
The observed proportion of hourly workers is approximately 24.18%.
The test gives:
- z-statistic: -3.5100
- p-value: 0.000448
- α: 0.05
Since p < 0.05, we reject H₀.

Conclusion
There is statistically significant evidence that the proportion of hourly workers is different from 25%.
The observed proportion is about 24.18%, which is slightly below 25%. With such a large sample, even this relatively small difference is statistically detectable.

### Final Summary
Challenge	Test - Main result	Decision
1. Explore data	Descriptive analysis	33,183 employees; 8,022 hourly	—
2. Mean hourly wage	One-sample t-test	t ≈ 20.62, p ≪ 0.05	Reject H₀
3. Mean hourly wage	95% CI	≈ $32.52–$33.05	$30 is outside CI
4. Hourly-worker proportion	One-sample proportion z-test	z ≈ −3.51, p ≈ 0.000448	Reject H₀


Overall conclusion
Both inferential tests provide statistically significant evidence against their respective null hypotheses:
1. The mean hourly wage is not $30/hour; the data indicate a higher mean of about $32.79/hour.
2. The proportion of employees who are hourly workers is not 25%; it is about 24.18%.

A final statistical reminder
Statistical significance does not automatically mean practical significance. The proportion test, for example, detects a difference of less than one percentage point because the dataset is very large. Always consider the size and practical importance of an effect, not only its p-value.